Test that newer libobjcs let the Swift runtime set up a class's layout

jrose-apple opened the pull request:
Specifically, when the class is referenced from Objective-C first, before being referenced from Swift in any way. This is important for resilience, since up until now Objective-C isn't used to the size of a class changing based on anything other than a superclass's size changing.

This is basically the test that would have gone with #19944 had there been a build of libobjc to test it with.

rdar://problem/45718008

jrose-apple commented:
Either of you see anything else worth testing while I'm here?

jrose-apple commented on validation-test/Runtime/class-layout-from-objc.m:
```diff
@@ -0,0 +1,72 @@
+
+  size_t offsetOfLast = ivar_getOffset(class_getInstanceVariable(c, "last"));
+  NSLog(@"offset of 'last' %zd", offsetOfLast);
+  assert(offsetOfLast == actualSize - sizeof(int64_t));
```
Whoops, I changed this last-minute.

jrose-apple commented:
@swift-ci Please test

slavapestov commented on validation-test/Runtime/Inputs/class-layout-from-objc/Classes.swift:
```diff
@@ -0,0 +1,40 @@
+  @objc var first: Int = 0
+  var forceUnalignment: Int8 = 0
+  var middle = GrowsToInt64()
+  @objc var last: Int = 0
```
Why are these (non-public) ivars @objc?

jrose-apple replied:
These are the ones I'm testing the ivar offsets of. It doesn't really matter if they're public or not; Swift never exposes direct compile-time ivar access to Objective-C.

jrose-apple replied:
If we ever *don't* emit them for non-`@objc` ivars, I don't want the test to change. I think it's appropriate for something accessed through the ObjC runtime. (Besides, the non-`@objc` ivars aren't representable in Objective-C, and it feels weird to rely on them being there.)

slavapestov commented on validation-test/Runtime/class-layout-from-objc.m:
```diff
@@ -0,0 +1,72 @@
+
+  size_t offsetOfLast = ivar_getOffset(class_getInstanceVariable(c, "last"));
+  NSLog(@"offset of 'last' %zd", offsetOfLast);
+  assert(offsetOfLast == actualSize - sizeof(intptr_t));
```
Is it worth creating an instance and checking anything there as well, maybe calling a Swift method too?

jrose-apple replied:
I couldn't think of anything that wouldn't go through a Swift entry point except `class_createInstance`, and I think that's superfluous with checking the instance size.

jrose-apple replied:
Well, we call static methods. Think an instance method is interesting enough?

slavapestov replied:
Probably not.

slavapestov commented on validation-test/Runtime/class-layout-from-objc.m:
```diff
@@ -0,0 +1,72 @@
+
+  size_t offsetOfFirst = ivar_getOffset(class_getInstanceVariable(c, "first"));
+  NSLog(@"offset of 'first' %zd", offsetOfFirst);
+  assert(offsetOfFirst == sizeof(Class) + (isPureSwift ? sizeof(void*) : 0));
```
Does this still hold on 32-bit? (I guess we'll find out)

jrose-apple replied:
Yeah, this is "non-pointer isa" or "isa + refcount". I checked on a 32-bit simulator.

slavapestov commented on validation-test/Runtime/Inputs/class-layout-from-objc/Classes.swift:
```diff
@@ -0,0 +1,40 @@
+  }
+}
+
+public class DynamicClass: NSObject {
```
Can you also test a subclass of an Objective-C class that is not NSObject (and has ivars)? We've had bugs with ivar layout in this case in the past.

jrose-apple replied:
Good idea, will add.

slavapestov commented:
This looks good. Thanks for doing this!

mikeash commented on validation-test/Runtime/Inputs/class-layout-from-objc/Resilient.swift:
```diff
@@ -0,0 +1,11 @@
+  #elseif BIG
+  var value: Int64
+  #else
+  #error("Must define OLD or NEW")
```
Should say SMALL or BIG?

jrose-apple replied:
Whoops, thanks!

mikeash reviewed (approved):
Looks good.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Looks like I got the 32-bit stuff wrong after my second version. Fortunately my 32-bit simulator is working now…!

jrose-apple committed:
Test that newer libobjcs let the Swift runtime set up a class's layout

Specifically, when the class is referenced from Objective-C first,
before being referenced from Swift in any way. This is important for
resilience, since up until now Objective-C isn't used to the size of a
class changing based on anything other than a superclass's size
changing.

This is basically the test that would have gone with 9024768b0 had
there been a build of libobjc to test it with.

rdar://problem/45718008

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux
